src/coreclr/jit/fgopt.cpp

```diff
@@ -7030,10 +7030,51 @@ PhaseStatus Compiler::fgHeadTailMerge(bool early)
         iterateTailMerge(retryBlocks.Pop());
     }
 
+    BasicBlock* returnTrueBb  = nullptr;
+    BasicBlock* returnFalseBb = nullptr;
+
     // Visit each block and try to merge first statements of successors.
     //
     for (BasicBlock* const block : Blocks())
     {
+        // Also, merge "return true/false" blocks regardless of where they are in the graph (but in the same EH region).
+        // Technically, we can generalize this algorithm to merge any "return CNS" blocks, but it seems to be profitable
+        // only for booleans (might be also profitable for 'return null;').
```

Comment:
> it seems to be profitable only for booleans 

Hmm, why is that?

The transformation is really a form of tail merging, so it seems odd that we do it during this loop that handles head merging. The tail merging loop above has a bunch of smart stuff in it to decide how to merge tails -- doesn't it help for this case too? What if you just build the normal `predInfo` for return blocks and then run the normal tail merge with it?

Reply:
@AndyAyersMS was hinting last week that he might do this work (https://github.com/dotnet/runtime/pull/94130#issuecomment-1783854877), so he probably has some thoughts too.

Reply:
> Hmm, why is that? 

Tried to allow it for `return null;` but only got a few diffs, I can try to generalize the algorithm to be `HashSet<BasicBlock>` + support multiple statements (split them) but I fear that will be a lot more code + TP impact with  small wins on top of this.

> The transformation is really a form of tail merging, so it seems odd that we do it during this loop that handles head merging.

I just needed some loop over blocks to check the diffs

Reply:
> I can try to generalize the algorithm to be HashSet<BasicBlock> + support multiple statements (split them) but I fear that will be a lot more code

Shouldn't the logic be basically the same as what is already there for tail merging already? Ideally we generalize that logic to handle returns instead of adding another separate implementation for it. The current implementation relies on a successor block to have a predecessor list. I think to generalize it we just build the `predInfo` directly for all return blocks (as if they had a pseudo successor).

I suppose the thing with return blocks is that we aren't going to have natural preexisting jumps to redirect, so there are additional heuristics necessary to determine profitability in the general case. I can see that the existing logic already has a preference towards picking blocks such they [won't need to be split](https://github.com/dotnet/runtime/blob/e733539a38458c5c3a64688b55a702444e470700/src/coreclr/jit/fgopt.cpp#L6900). That seems like one of the primary things in the implementation of this PR, so maybe require it for the new `BBJ_RETURN` case if that's the necessary profitability heuristic?

What I'm really after is making sure that we unify things to not end up with two different code paths that try to do basically the same thing.

Reply:
thanks! Will take a look!

Reply:
My main motivation is to unconditionally expand `return <cond>;` early to `JTRUE ret true/false` so then other opts like "optimize bool" can only focus on JTRUE (sort of cannonical form for all branchless opts, etc) and the redundant branch opt can benefit from this too.

Then, this opt merges all `return true/false` tails and then some other phases merges
```
JTRUE
return false
return true
```
back to RETURN if no tails were merged.

Reply:
I think in the long run we should consider merging all returns early (save perhaps for tail call candidates) and keep things merged until quite late, since we don't know the actual epilog costs until after LSRA.

I also wonder if there is value in generalizing this sort of thing, say any time we assign the result of a relop to a local at the end of a non-branching block that targets a join (where the return successor is an implicit kind of join), we do this expansion and then let tail merge see if it can clean things up....

Reply:
That would also expose the predicates to RBO... it currently doesn't do much for predicates that don't feed branches.